### PR TITLE
 stats: fix row estimation for pseudo unique key

### PR DESCRIPTION
**Pull request by lamxTyler**
If there are equal conditions on the unique key, the row count should not be greater than 1.
PTAL @coocood @winoros

### CODE REVIEW

File: statistics/table.go [hunk 1]
```diff
@@ -325,26 +325,41 @@ func (t *Table) GetRowCountByColumnRanges(sc *stmtctx.StatementContext, colID in
...
 	result, err := idx.getRowCount(sc, indexRanges)
 	result *= idx.getIncreaseFactor(t.Count)
 	return result, errors.Trace(err)
 }
 
 // PseudoTable creates a pseudo table statistics.
-func PseudoTable(tableID int64) *Table {
-	return &Table{
-		TableID: tableID,
+func PseudoTable(tblInfo *model.TableInfo) *Table {
```
Comment (shenli) [inline statistics/table.go]: Could we cache the pseudo table stats in TableInfo?

File: statistics/selectivity.go [hunk 2]
```diff
@@ -44,32 +44,66 @@ const (
...
+		fun, ok := expr.(*expression.ScalarFunction)
+		if !ok {
+			continue
+		}
+		colID := getConstantColumnID(fun.GetArgs())
+		if colID == -1 {
+			continue
+		}
+		switch fun.FuncName.L {
+		case ast.EQ, ast.NullEQ, ast.In:
```
Comment (coocood) [inline statistics/selectivity.go]: `In` and `EQ` has the same selectivity?

File: statistics/selectivity.go [hunk 3]
```diff
@@ -44,32 +44,66 @@ const (
...
+		}
+		unique := true
+		for _, col := range idx.Info.Columns {
+			if !uniqueCol[col.Name.L] {
+				unique = false
+				break
+			}
+		}
+		if unique {
+			return 1.0 / float64(t.Count)
```
Comment (coocood) [inline statistics/selectivity.go]: What if the expression is `IN` with many unique values?

File: statistics/selectivity.go [hunk 4]
```diff
@@ -44,32 +44,66 @@ const (
...
-			case ast.EQ, ast.NullEQ:
-				minFactor = math.Min(minFactor, 1.0/pseudoEqualRate)
-			case ast.GE, ast.GT, ast.LE, ast.LT:
-				minFactor = math.Min(minFactor, 1.0/pseudoLessRate)
-				// FIXME: To resolve the between case.
+		fun, ok := expr.(*expression.ScalarFunction)
+		if !ok {
+			continue
+		}
+		colID := getConstantColumnID(fun.GetArgs())
```
Comment (coocood) [inline statistics/selectivity.go]: Should get ColumnID in expressions like `c in (1, 2, 3)`?

File: statistics/selectivity.go [hunk 5]
```diff
@@ -44,32 +44,66 @@ const (
...
 )
 
-// checkColumnConstant receives two expressions and makes sure one of them is column and another is constant.
-func checkColumnConstant(e []expression.Expression) bool {
+// getConstantColumnID receives two expressions and if one of them is column and another is constant, it returns the
+// ID of the column.
+func getConstantColumnID(e []expression.Expression) int64 {
 	if len(e) != 2 {
-		return false
+		return -1
```
Comment (coocood) [inline statistics/selectivity.go]: Should get ColumnID in expressions like c in (1, 2, 3)?

File: statistics/selectivity.go (hunk 2 above)
Comment (zz-jason) [reply statistics/selectivity.go]: `getConstantColumnID` guarantees there exists only two parameters for the current expr, so `in` is equivalent to `eq`

File: statistics/table.go (hunk 1 above)
Comment (lamxTyler) [reply statistics/table.go]: Yes.

File: statistics/handle.go [hunk 6]
```diff
@@ -145,10 +146,12 @@ func (h *Handle) Update(is infoschema.InfoSchema) error {
...
 
 // GetTableStats retrieves the statistics table from cache, and the cache will be updated by a goroutine.
-func (h *Handle) GetTableStats(tblID int64) *Table {
-	tbl, ok := h.statsCache.Load().(statsCache)[tblID]
+func (h *Handle) GetTableStats(tblInfo *model.TableInfo) *Table {
+	tbl, ok := h.statsCache.Load().(statsCache)[tblInfo.ID]
 	if !ok {
-		return PseudoTable(tblID)
+		tbl = PseudoTable(tblInfo)
+		h.UpdateTableStats([]*Table{tbl}, nil)
```
Comment (coocood) [inline statistics/handle.go]: Can we not mix pseudo table with non-pseudo ones by adding `pseudoCache` in `Handle`?

Comment (lamxTyler) [reply statistics/handle.go]: What's the benefit?

File: statistics/handle.go [hunk 7]
```diff
@@ -179,7 +182,10 @@ func (h *Handle) UpdateTableStats(tables []*Table, deletedIDs []int64) {
 func (h *Handle) LoadNeededHistograms() error {
 	cols := histogramNeededColumns.allCols()
 	for _, col := range cols {
-		tbl := h.GetTableStats(col.tableID).copy()
+		tbl, ok := h.statsCache.Load().(statsCache)[col.tableID]
```
Comment (coocood) [inline statistics/handle.go]: Why not copy the tbl?

File: statistics/handle.go (hunk 6 above)
Comment (coocood) [reply statistics/handle.go]: So we don't need to load histograms for pseudo ones.

Comment (lamxTyler) [reply statistics/handle.go]: But it won't load histograms for pseudo ones.

**Pull request merged**